File: app/src/main/java/ru/laushkina/rates/repository/RatesDbRepository.java
package ru.laushkina.rates.repository;

import android.content.Context;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Maybe;
import io.reactivex.Single;
import ru.laushkina.rates.model.Rate;

public class RatesDbRepository implements RatesRepository {
    private final RatesDatabase ratesDatabase;

    public RatesDbRepository(@NonNull Context context) {
        ratesDatabase = RatesDatabase.getInstance(context);
    }

    @Override
    public void save(@NonNull List<Rate> rates) {
        ratesDatabase.getTransactionExecutor().execute(() -> {
            this.ratesDatabase.rateDao().truncate();
            this.ratesDatabase.rateDao().save(mapToEntity(rates));
        });
    }

    @NonNull
    @Override
    public Single<List<Rate>> getRates() {
        return ratesDatabase.rateDao()
                .loadAll()
                .map(RatesDbRepository::mapToRate);
    }

    @NonNull
    @Override
    public Maybe<Rate> getBaseRate() {
        return ratesDatabase.rateDao()
                .loadBaseRate()
                .map(entity -> new Rate(entity.shortName, entity.amount, entity.isBase));
    }

    @NonNull
    private static List<RateEntity> mapToEntity(@NonNull List<Rate> rates) {
        List<RateEntity> result = new ArrayList<>(rates.size());
        for (Rate rate : rates) {
            result.add(new RateEntity(rate));
        }
        return result;
    }

    @NonNull
    private static List<Rate> mapToRate(@NonNull List<RateEntity> entities) {
        List<Rate> result = new ArrayList<>(entities.size());
        for (RateEntity entity : entities) {
            result.add(new Rate(entity.shortName, entity.amount, entity.isBase));
        }
        return result;
    }
}

File: app/src/main/java/ru/laushkina/rates/repository/RatesDbRepository.kt
package ru.laushkina.rates.repository

import android.content.Context
import io.reactivex.Maybe
import ru.laushkina.rates.model.Rate
import ru.laushkina.rates.model.RateShortName.Companion.parse
import java.util.*

//TODO datasource
class RatesDbRepository(context: Context): RatesRepository {
    private val ratesDatabase: RatesDatabase = RatesDatabase.getInstance(context)

    override fun save(rates: List<Rate>?) {
        ratesDatabase.transactionExecutor.execute {
            ratesDatabase.rateDao().truncate()
            ratesDatabase.rateDao().save(mapToEntity(rates))
        }
    }

    override fun getRates(): Maybe<List<Rate>> {
        return ratesDatabase.rateDao()
                .loadAll()
                .map { entities: List<RateEntity> -> mapToRate(entities) }
    }

    override fun getBaseRate(): Maybe<Rate> {
        return ratesDatabase.rateDao()
                .loadBaseRate()
                .map { entity -> Rate(parse(entity.shortName), entity.amount, entity.isBase) }
    }

    private fun mapToEntity(rates: List<Rate>?): List<RateEntity>? {
        if (rates == null) return null

        val result: MutableList<RateEntity> = ArrayList(rates.size)
        for (rate in rates) {
            result.add(RateEntity(rate))
        }
        return result
    }

    private fun mapToRate(entities: List<RateEntity>): List<Rate> {
        val result: MutableList<Rate> = ArrayList(entities.size)
        for (entity in entities) {
            result.add(Rate(parse(entity.shortName), entity.amount, entity.isBase))
        }
        return result
    }
}